test/tests/functional/pbs_conf_resv_stale_vnode.py

```diff
@@ -0,0 +1,116 @@
+# Public License agreement ("AGPL"), except where a separate commercial license
+# agreement for PBS Pro version 14 or later has been executed in writing with
+# Altair.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under
+# a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+
+from tests.functional import *
+
+
+class ResvStaleVnode(TestFunctional):
+    """
+    Test that the scheduler won't confirm a reservation on stale vnode and
+    make sure reservations that have nodes that have gone stale get degreaded
+    """
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        # Create 3 vnodes named different things in different vnodedef files
+        # This allows us to delete a vnodedef file and make that node stale
+        self.mom.add_config(conf={'$vnodedef_additive': 'False'})
+        a = {'resources_available.ncpus': 1}
+        self.server.create_vnodes('foo', a, 1, fname='nat', restart=False,
+                                  mom=self.mom, usenatvnode=True, expect=False)
+        self.server.create_vnodes('vn', a, 1, fname='fname1', delall=False,
+                                  restart=False, additive=True, mom=self.mom,
+                                  expect=False)
+        self.server.create_vnodes('vnode', a, 1, fname='fname2', delall=False,
+                                  additive=True, mom=self.mom, expect=False)
+
+    def test_conf_resv_stale_vnode(self):
+        """
+        Test that the scheduler won't confirm a reservation on a stale node.
+        """
+        # Ensure the hostsets aren't used by associating a node to a queue
+        a = {'queue_type': 'Execution', 'enabled': 'True', 'started': 'True'}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, id='workq2')
+        self.server.manager(MGR_CMD_SET, NODE, {'queue': 'workq2'},
+                            id=self.mom.shortname)
+
+        # Submit a job that will run on our stale vnode
+        a = {'Resource_List.select': '1:vnode=vn[0]',
+             'Resource_List.walltime': 3600}
+        J = Job(TEST_USER, attrs=a)
+        jid = self.server.submit(J)
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+
+        self.mom.delete_vnode_defs(vdefname='fname1')
+        self.mom.signal('-HUP')
+        self.server.expect(NODE, {'state': (MATCH_RE, 'Stale')}, id='vn[0]')
+
+        now = int(time.time())
+        a = {'reserve_start': now + 5400, 'reserve_end': now + 7200}
+        R = Reservation(TEST_USER, a)
+        rid = self.server.submit(R)
+
+        # Reservation should be confirmed on vnode[0] since vn[0] is Stale
+        a = {'resv_nodes': '(vnode[0]:ncpus=1)'}
```

Comment:
is this result dependent on the fact that scheduler always see vn[0] first in the list of vnodes? I don't see a way to proof that vn[0] was ignored because of its stale status